Tools/cases_generator/generate_cases.py

```diff
@@ -236,14 +284,20 @@ def write(self, out: Formatter) -> None:
 
         if not self.register:
             # Write net stack growth/shrinkage
-            diff = len(self.output_effects) - len(self.input_effects)
-            out.stack_adjust(diff)
+            out.stack_adjust(0, self.input_effects, self.output_effects)
 
             # Write output stack effect assignments
-            for i, oeffect in enumerate(reversed(self.output_effects), 1):
-                if oeffect.name not in self.unmoved_names:
-                    dst = StackEffect(f"PEEK({i})", "")
-                    out.assign(dst, oeffect)
+            oeffects = list(reversed(self.output_effects))
+            for i in range(len(oeffects)):
+                oeffect = oeffects[i]
+                if oeffect.name in self.unmoved_names:
+                    continue
+                osize = string_effect_size(list_effect_size(oeffects[:i+1]))
+                if oeffect.size:
+                    dst = StackEffect(f"&PEEK({osize})", "PyObject **")
+                else:
+                    dst = StackEffect(f"PEEK({osize})", "")
```

Comment:
It seems like this whole block is identical to the logic for input effects. Maybe factor out into a function?

Reply:
I'm not excited about having to pick a name for a helper method that's five lines and takes three parameters (i, ieffect, ieffects).

Reply:
Ah, I missed the differences at the top and bottom of the for loops (I thought the entire loop could be factored out). No need to do this then.